FSW/src/util/faultManager.cpp

```diff
@@ -0,0 +1,303 @@
+/* faultManager.cpp handles NS2 payload fault mitigation
+ * Usage:
+ *  module functionality
+ *  function definitions
+ *  put function declarations in memUtil.hpp
+ * 
+ * Modules encompassed:
+ *  Fault Detection
+ *  Fault Logging
+ *  Fault Mitigation
+ *
+ * Additional files needed for compilation:
+ *  config.hpp
+ *  encodedFile.hpp
+ *  
+ */
+
+/* - - - - - - Includes - - - - - - */
+// All libraries are put in dataCollection.hpp
+// NS2 headers
+#include "../headers/faultManager.hpp"
+#include "../headers/encodedFile.hpp"
+#include "../headers/commandHandling.hpp"
+#include "../headers/housekeeping.hpp"
+
+/* Module Variable Definitions */
+// fault log
+static FaultReport faultLog[faultCode::COUNT];
+// decoded size of data on EEPROM, in terms of bytes
+const size_t EEPROM_DECODED_SIZE = faultCode::COUNT * FaultReport::MEMSIZE + PayloadData::MEMSIZE; 
+// flag to indicate whether any faults were logged in the current main loop iteration
+static bool detectedFault = false; 
+
+/* - - - - - - Module Driver Functions - - - - - - */
+
+/* - - - - - - logFault - - - - - - *
+ * Usage:
+ *  records a new occurence of a fault
+ *  
+ * Inputs:
+ *  code - fault code to log
+ * 
+ * Outputs:
+ *  None
+ */
+void logFault(uint8_t code) {
+    detectedFault = true;
+
+    // check if fault is valid
+    if (code >= faultCode::ERR_CODE) {
+        Serial.println("Warning, tried to log a fault with invalid fault code (Fault Manager)");
+        return;
+    }
+
+    Serial.print("Fault Logged: ");
+    Serial.println(code);
+
+    // update corresponding fault report
+    faultLog[code].startNum = payloadData.startCount;
+    faultLog[code].timestamp = millis();
+
+    if (faultLog[code].occurences < 255) {
+        faultLog[code].occurences++;
+        
+        // mark for action on first and max occurences
+        if (faultLog[code].occurences == 1 || faultLog[code].occurences == 255) {
+            faultLog[code].pendingAction = 1;
+        }
+    }
+}
+
+/* - - - - - - feedWD - - - - - - *
+ * Usage:
+ *  resets the watchdog
+ *  
+ * Inputs:
+ *  None
+ * Outputs:
+ *  None
+ */
+void feedWD() {
+    // TODO: some checks here to make sure everything is okay before feeding the dog?
```

Comment:
I can't think of any. I suppose you could check that the output of the voltage regulator is good, but if the voltage regulator stops outputting then the Teensy wouldn't be powered anyway. You could check the temperatures, and not send the watchdog signal if the temperature goes too high or low, but then the issue is that if the Teensy doesn't restart and start sending the signal again you can't ever turn it back on. 

I think some more work needs to be done to think of a better watchdog solution, so I wouldn't worry about it for now.

Reply:
I agree, I'll leave it like that for now. The worst case scenario is still that the software gets caught up indefinitely, which is the point of having a watchdog at all.